fluent/src/bundle.js

```diff
@@ -170,74 +161,80 @@ export default class FluentBundle {
    *
-   * Format a raw `message` from the bundle into a string (or a null if it has
-   * a null value).  `args` will be used to resolve references to variables
-   * passed as arguments to the translation.
+   * Format a raw `Pattern` into a string. `args` will be used to resolve
+   * references to variables passed as arguments to the translation.
    *
-   * In case of errors `format` will try to salvage as much of the translation
-   * as possible and will still return a string.  For performance reasons, the
-   * encountered errors are not returned but instead are appended to the
-   * `errors` array passed as the third argument.
+   * In case of errors `formatPattern` will try to salvage as much of the
+   * translation as possible and will still return a string. For performance
+   * reasons, the encountered errors are not returned but instead are appended
+   * to the `errors` array passed as the third argument.
    *
    *     const errors = [];
    *     bundle.addMessages('hello = Hello, { $name }!');
    *     const hello = bundle.getMessage('hello');
-   *     bundle.format(hello, { name: 'Jane' }, errors);
-   *
-   *     // Returns 'Hello, Jane!' and `errors` is empty.
-   *
-   *     bundle.format(hello, undefined, errors);
-   *
-   *     // Returns 'Hello, name!' and `errors` is now:
-   *
-   *     [<ReferenceError: Unknown variable: name>]
-   *
-   * @param   {Object | string}    message
-   * @param   {Object | undefined} args
-   * @param   {Array}              errors
-   * @returns {?string}
+   *     if (hello.value) {
+   *         bundle.formatPattern(hello.value, { name: 'Jane' }, errors);
+   *         // Returns 'Hello, Jane!' and `errors` is empty.
+   *
+   *         bundle.formatPattern(hello.value, undefined, errors);
+   *         // Returns 'Hello, {$name}!' and `errors` is now:
+   *         // [<ReferenceError: Unknown variable: name>]
+   *     }
+   *
+   * @param   {string|Array} pattern
+   * @param   {?Object} args
+   * @param   {?Array} errors
+   * @returns {string}
    */
-  format(message, args, errors) {
-    // optimize entities which are simple strings with no attributes
-    if (typeof message === "string") {
-      return this._transform(message);
+  formatPattern(pattern, args, errors) {
+    // Resolve a simple pattern without creating a scope.
+    if (typeof pattern === "string") {
+      return this._transform(pattern);
     }
 
-    // optimize entities with null values
-    if (message === null || message.value === null) {
-      return null;
+    // Resolve a complex pattern.
+    if (Array.isArray(pattern)) {
+      let scope = this._createScope(args, errors);
+      let value = resolveComplexPattern(scope, pattern);
+      return value.toString(scope);
     }
 
-    // optimize simple-string entities with attributes
-    if (typeof message.value === "string") {
-      return this._transform(message.value);
-    }
+    throw new TypeError("Invalid Pattern type");
```

Comment:
I'm not really sure what's the right thing to do here. Should we guard against unexpected types of `patterns`? This could be helpful when someone forgets to check if `message.value` is not `null` before calling `formatPattern`. If I don't throw here, `resolveComplexPattern` likely will, anyways, but with a more cryptic error.

Reply:
My rationale for throwing here is that this isn't a formatting errors. It's not caused by a bug in the translation which resulted in an incompletely formatted output. As such, I don't think it fits into `errors` either. This is an error in how the API has been used by the consumer. If the consumer takes care to always only pass `Patterns` to `formatPattern`, this error will never happen.

Reply:
I guess your last comment also addresses your initial concern if developers should guard their calls with `try catch` ?

Also, we're not really protecting this API by checking if the input is an `Array` against being fed garbage, `[{foopy:true}]` is still bad ;-)

If `Pattern` was a subclass of `Array`, we might have a slightly stronger assertion?

Reply:
> I guess your last comment also addresses your initial concern if developers should guard their calls with try catch ?

Oh! Yes, I think so! :)

Let me try to rephrase this to make sure I understand your point: If I'm fine expecting developers to use a `try catch`, I should also be fine expecting them to check `if (message.value)` before calling `formatPattern`. Is this what you meant?

Reply:
I'm not sure what you're trying to achieve. If this is just about `message.value` being null unexpectedly, I'd consider `null` to be actually an OK return value.

I read this to be about creating a defined error scenario when people pass in garbage like `Math.random` or so. I wouldn't expect good users of the API to `try-catch` to proof them for API abuse.

Reply:
OK, thanks for clarifying. I'd like to consider adding `null` as an OK return value in a follow-up. I'd prefer if this method only returned strings, so that it's not necessary to check for non-null output after every `formatPattern`.

For now, I'll keep the throwing on non-string, non-array pattern in order to help developers find buggy callsites faster.